### PR TITLE
Add factories for SamplingResult

### LINKED ISSUE #3354: Add factories for SamplingResult
We can currently only implement a Sampler by calling `SamplingResult.create`. For results without attributes, we do return singletons from this method so it doesn't have any real overhead. But it's somewhat annoying to have to call `create` with a parameter and not so intuitive - every time I do it I still ctrl-click through the method to confirm it's not creating a new object. The alternative of creating a `private static final` in the Sampler seems redundant vs us providing some factories

`SamplingResult.recordAndSample()`
`SamplingResult.recordOnly()`
`SamplingResult.drop()`

### PR DESCRIPTION
**Pull request by Rocksnake**
https://github.com/open-telemetry/opentelemetry-java/pull/3360

Fixes #3354

rebase it by re-create the branch and re-push

### CODE REVIEW

**Review by anuraaga — APPROVED** [review]
Thanks!

**Review by kubawach — APPROVED** [review]
LGTM :)

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java [hunk 1]
```diff
@@ -67,6 +67,45 @@ static SamplingResult create(SamplingDecision decision, Attributes attributes) {
...
   }
 
+  /**
+   * Returns a {@link SamplingResult} corresponding to {@link SamplingDecision#RECORD_AND_SAMPLE}
+   * with no attributes and {@link SamplingResult#getDecision()} returning {@code decision}.
+   *
+   * <p>This is meant for use by custom {@link Sampler} implementations and is equivalent to calling
+   * {@code SamplingResult.create(SamplingDecision.RECORD_AND_SAMPLE)}.
+   *
+   * @return A {@link SamplingResult} with empty attributes and the provided {@code decision}.
```
Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:77]: This javadoc isn't correct. The user isn't providing a decision with this call.

Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:77]:
```suggestion
   * @return A "record and sample" {@link SamplingResult} with empty attributes.
```

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java [hunk 2]
```diff
@@ -67,6 +67,45 @@ static SamplingResult create(SamplingDecision decision, Attributes attributes) {
         : ImmutableSamplingResult.createSamplingResult(decision, attributes);
   }
 
+  /**
+   * Returns a {@link SamplingResult} corresponding to {@link SamplingDecision#RECORD_AND_SAMPLE}
+   * with no attributes and {@link SamplingResult#getDecision()} returning {@code decision}.
```
Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:72]:
```suggestion
   * with no attributes.
```

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java [hunk 3]
```diff
@@ -67,6 +67,45 @@ static SamplingResult create(SamplingDecision decision, Attributes attributes) {
...
+   *
+   * @return A {@link SamplingResult} with empty attributes and the provided {@code decision}.
+   */
+  static SamplingResult recordAndSample() {
+    return ImmutableSamplingResult.EMPTY_RECORDED_AND_SAMPLED_SAMPLING_RESULT;
+  }
+
+  /**
+   * Returns a {@link SamplingResult} corresponding to {@link SamplingDecision#RECORD_ONLY} with no
+   * attributes and {@link SamplingResult#getDecision()} returning {@code decision}.
```
Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:85]:
```suggestion
   * attributes.
```

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java [hunk 4]
```diff
@@ -67,6 +67,45 @@ static SamplingResult create(SamplingDecision decision, Attributes attributes) {
...
+   *
+   * <p>This is meant for use by custom {@link Sampler} implementations and is equivalent to calling
+   * {@code SamplingResult.create(SamplingDecision.RECORD_ONLY)}.
+   *
+   * @return A {@link SamplingResult} with empty attributes and the provided {@code decision}.
```
Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:90]:
```suggestion
   * @return A "record only" {@link SamplingResult} with empty attributes.
```

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java [hunk 5]
```diff
@@ -67,6 +67,45 @@ static SamplingResult create(SamplingDecision decision, Attributes attributes) {
...
+   */
+  static SamplingResult recordOnly() {
+    return ImmutableSamplingResult.EMPTY_RECORDED_SAMPLING_RESULT;
+  }
+
+  /**
+   * Returns a {@link SamplingResult} corresponding to {@link SamplingDecision#DROP} with no
+   * attributes and {@link SamplingResult#getDecision()} returning {@code decision}.
```
Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:98]:
```suggestion
   * attributes.
```

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java [hunk 6]
```diff
@@ -67,6 +67,45 @@ static SamplingResult create(SamplingDecision decision, Attributes attributes) {
...
+   *
+   * <p>This is meant for use by custom {@link Sampler} implementations and is equivalent to calling
+   * {@code SamplingResult.create(SamplingDecision.DROP)}.
+   *
+   * @return A {@link SamplingResult} with empty attributes and the provided {@code decision}.
```
Comment (jkwatson) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java:103]:
```suggestion
   * @return A "drop" {@link SamplingResult} with empty attributes.
```

**Pull request merged**